### DISCUSSION TITLE
Authenticating git command line with personal access token

### DISCUSSION THREAD

**Discussion by munnellg** [community/community#23419] [New to GitHub]
<p>Hi there,</p>
<p>I got an email today while pushing an update to a repo to say that password authentication will be deprecated later this year in favour of access tokens. I want to set myself up for this change now rather than waiting until August. Better fix now rather than later :slight_smile:</p>
<p>I read the notice linked to by the email, and the documentation linked from there, and went about setting up a personal access token:</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/" target="_blank" rel="noopener nofollow ugc" title="08:28PM - 15 December 2020">The GitHub Blog – 15 Dec 20</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/09/security-1200-630.png?fit=1200%2C630" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/" target="_blank" rel="noopener nofollow ugc">Token authentication requirements for Git operations - The GitHub Blog</a></h3>

<p>Beginning August 13th, 2021, we will no longer accept account passwords when authenticating Git operations on GitHub.com and will instead require token-based authentication (for example, a personal access, OAuth, or GitHub App installation token) for...</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>However, I’m a little confounded by the last step which seems to suggest that I have to copy the PAT into the password field for git whenever I make a request?</p>
<p>Quote from the documentation below:</p>
<blockquote>
<p>Once you have a token, you can enter it instead of your password when performing Git operations over HTTPS.</p>
<p>For example, on the command line you would enter the following:</p>
</blockquote>
<pre><code class="lang-auto">$ git clone https://github.com/username/repo.git
Username: your_username
Password: *your_token*
</code></pre>
<p>Is this really the only way to use a PAT? Surely there must be a way to authenticate without my having to copy/paste the token from the clipboard every time I want to push a commit? Moreover, the github website only shows the token to you once when it is created. So do I have to store the token in a plain text file on my machine?</p>
<p>Some guidance here would be much appreciated because I’m quite sure I’ve missed something. I had imagined that there would be a way to use a PAT like a public/private key where I would save it in a secure file on my system, could inform git of the location of the file and then git would handle all necessary authentication automatically in repsonse to requests.</p>
<p>Many thanks for your time!</p>

**Comment by canuckjacq** [comment, marked as answer]
<p>:wave: Welcome!</p>
<p>A personal access token is used exactly like a password. That means you can <a href="https://docs.github.com/en/github/getting-started-with-github/caching-your-github-credentials-in-git">cache your token in git</a>, or a password manager.</p>
<p>More here about git credential storage: <a href="https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage" class="inline-onebox">Git - Credential Storage</a>.</p>

**Comment by munnellg** [comment]
<p>Hi canuckjacq,</p>
<p>Thanks for the info, that’s good to know! I’ll experiment with git store for now. It seems like the most natural solution.</p>
<p>I think it would be good if a few extracts from the resources you linked could be added to the article below. If github is moving to PAT based authentication then this is likely where people such as myself will get their information. It would be helpful to know what options we have to spare ourselves copy/pasting over and over again.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" target="_blank" rel="noopener nofollow ugc">Creating a personal access token - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by canuckjacq** [comment]
<p>Thanks for your feedback! I’ll open an issue - to be honest I found our documentation a little confusing myself, which is why I linked to the external page.</p>

**Comment by munnellg** [comment]
<p>Awesome! I actually only just noticed that the docs are open source anyway. I might make the edit myself and open a pull request. The page isn’t too bad, to be fair. Just that bit at the end could use some work.</p>

**Comment by mintty** [comment]
<p>I have no idea why github is bothering us with changing passwords to credentials that are supposed to work in the same way. Even more it is extremely annoying that they do not provide any <em>short</em> how-to instructions to get the credentials deployed.<br>
Pointing to some lenghty elaborations about credentials and tools is not a response to this issue. It’s wasting time of many people who are enforced to study things they don’t care about. Please be more respectful to your users.<br>
Please provide instructions what to do once those credentials are created. Just the commands, not any theory.</p>

**Comment by jsoref** [comment]
<p>Passwords and tokens may fit into the same boxes, but they aren’t the same.</p>
<p>A GitHub PAT can’t be used to log into your github account and change settings.</p>
<p>And you can disable a PAT w/o breaking your account.</p>
<p>Also, GitHub can recognize when you’ve leaked a PAT and disable it, w/o blocking your account.</p>
<p>These are all valuable properties.</p>
<p>In general, I’d suggest setting up SSH keys.</p>
<p>One thing GitHub may hopefully eventually do is force web based logins to use 2FA, and you can’t fit a 2FA challenge-response thing into a simple https url that you give to a git client. So by divorcing the git https credential from the web login credential, it enables GitHub to eventually move users to 2FA w/o breaking their ability to use https for git operations.</p>
<p><a href="https://docs.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa/accessing-github-using-two-factor-authentication" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa/accessing-github-using-two-factor-authentication</a></p>

**Comment by jsoref** [comment]
<p>Further to the point, PATs allow you to scope permissions, instead of a password which enables the bearer to do <em>everything</em> as the user, a PAT can be scoped to perform a limited set of operations. This is incredibly valuable:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://www.beyondtrust.com/blog/entry/what-is-least-privilege">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/2119212/181076844-47aa95e9-f6d2-454b-92ca-1fe217251ce9.png" class="site-icon" width="192" height="192">

      <a href="https://www.beyondtrust.com/blog/entry/what-is-least-privilege" target="_blank" rel="noopener nofollow ugc">beyondtrust.com</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181076845-97caad1d-66d6-4fd1-8740-ee26462cd98f.jpeg" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c06745cc1eaf1826b31f24e2cc78c031d05030d1_2_690x345.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c06745cc1eaf1826b31f24e2cc78c031d05030d1_2_1035x517.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c06745cc1eaf1826b31f24e2cc78c031d05030d1_2_1380x690.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c06745cc1eaf1826b31f24e2cc78c031d05030d1_2_10x10.png"></div>

<h3><a href="https://www.beyondtrust.com/blog/entry/what-is-least-privilege" target="_blank" rel="noopener nofollow ugc">What Is Least Privilege &amp; Why Do You Need It? | BeyondTrust</a></h3>

  <p>Least privilege is the concept and practice of restricting access rights for users, accounts, and computing processes to only those resources absolutely required to perform routine, legitimate activities. This blog provides an in-depth overview of...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by mintty** [comment]
<p>Thanks for the comment but it does not help with the question, how do I deploy the new key instead of the previous password, so I won’t get those deprecation warnings anymore (which suggest my login wouldn’t work after the switch date).</p>

**Comment by airtower-luna** [comment]
<p>In principle: Use the token where you previously used your password.</p>
<p>The details vary depending on the client you use. Note that you may have to delete the password from a system keyring/password cache to be asked again, if your system has one.</p>

**Comment by mintty** [comment]
<p>OK, I use git command line only (in cygwin). No idea where it may have cached the credentials or how to delete/update them. But maybe it will ask implicitly once the password method has expired.</p>

**Comment by jsoref** [comment]
<aside class="onebox stackexchange" data-onebox-src="https://stackoverflow.com/questions/15381198/remove-credentials-from-git">
  <header class="source">

      <a href="https://stackoverflow.com/questions/15381198/remove-credentials-from-git" target="_blank" rel="noopener nofollow ugc">stackoverflow.com</a>
  </header>

  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/28004/balexandre" target="_blank" rel="noopener nofollow ugc">
    <img alt="balexandre" src="https://user-images.githubusercontent.com/2119212/181076849-5dd2475a-6c24-42ca-9b88-7e94fc49dc4e.png" class="thumbnail onebox-avatar" width="128" height="128">
  </a>

<h4>
  <a href="https://stackoverflow.com/questions/15381198/remove-credentials-from-git" target="_blank" rel="noopener nofollow ugc">Remove credentials from Git</a>
</h4>

<div class="tags">
  <strong>windows, git, credentials</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://stackoverflow.com/users/28004/balexandre" target="_blank" rel="noopener nofollow ugc">
    balexandre
  </a>
  on <a href="https://stackoverflow.com/questions/15381198/remove-credentials-from-git" target="_blank" rel="noopener nofollow ugc">09:20AM - 13 Mar 13 UTC</a>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
